homeassistant/components/asuswrt/__init__.py

```diff
@@ -1,120 +1,141 @@
+                    vol.Exclusive(CONF_PASSWORD, SECRET_GROUP): cv.string,
+                    vol.Exclusive(CONF_SSH_KEY, SECRET_GROUP): cv.isfile,
+                    vol.Exclusive(CONF_PUB_KEY, SECRET_GROUP): cv.isfile,
+                    vol.Optional(CONF_SENSORS): vol.All(
+                        cv.ensure_list, [vol.In(SENSOR_TYPES)]
+                    ),
+                    vol.Optional(CONF_INTERFACE, default=DEFAULT_INTERFACE): cv.string,
+                    vol.Optional(CONF_DNSMASQ, default=DEFAULT_DNSMASQ): cv.string,
+                }
+            )
+        },
+    ),
     extra=vol.ALLOW_EXTRA,
 )
 
 
-async def async_setup(hass, config, retry_delay=FIRST_RETRY_TIME):
-    """Set up the asuswrt component."""
+async def async_setup(hass, config):
+    """Set up the AsusWrt integration."""
+    conf = config.get(DOMAIN)
+    if conf is None:
+        return True
 
-    conf = config[DOMAIN]
+    domains_list = hass.config_entries.async_domains()
+    if DOMAIN in domains_list:
+        return True
 
-    api = AsusWrt(
-        conf[CONF_HOST],
-        conf[CONF_PORT],
-        conf[CONF_PROTOCOL] == "telnet",
-        conf[CONF_USERNAME],
-        conf.get(CONF_PASSWORD, ""),
-        conf.get("ssh_key", conf.get("pub_key", "")),
-        conf[CONF_MODE],
-        conf[CONF_REQUIRE_IP],
-        interface=conf[CONF_INTERFACE],
-        dnsmasq=conf[CONF_DNSMASQ],
+    pub_key = conf.get(CONF_PUB_KEY)
+    if pub_key:
+        conf[CONF_SSH_KEY] = pub_key
+    conf.pop(CONF_PUB_KEY, "")
+    conf.pop(CONF_SENSORS, {})
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_IMPORT}, data=conf
+        )
     )
 
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry):
+    """Set up AsusWrt platform."""
+
+    api = get_api(entry.data)
+
     try:
         await api.connection.async_connect()
-    except OSError as ex:
-        _LOGGER.warning(
-            "Error [%s] connecting %s to %s. Will retry in %s seconds...",
-            str(ex),
-            DOMAIN,
-            conf[CONF_HOST],
-            retry_delay,
-        )
+        if not api.is_connected:
+            raise ConfigEntryNotReady
+    except OSError as exp:
+        raise ConfigEntryNotReady from exp
 
-        async def retry_setup(now):
-            """Retry setup if a error happens on asuswrt API."""
-            await async_setup(
-                hass, config, retry_delay=min(2 * retry_delay, MAX_RETRY_TIME)
-            )
+    hass.data[DATA_ASUSWRT] = api
 
-        async_call_later(hass, retry_delay, retry_setup)
+    for platform in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, platform)
+        )
 
-        return True
+    async def async_close_connection(event):
+        """Close AsusWrt connection on HA Stop."""
+        if hasattr(api.connection, "disconnect"):
```

Comment:
Why is the api dynamic?

Reply:
Because `disconnect` method is available only for `telnet` and not `ssh`.
I thought it would be good manage disconnection when required and method is available.

Reply:
Do we know the type of connection from the config entry data? If so, I'd suggest to check that instead of using `hasattr`.